src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,288 @@
+    buffer.get(idArray)
+
+    val id = new String(idArray)
+    val workerAddressSize = buffer.getInt()
+    val workerAddress = buffer.slice()
+
+    workerAddress.limit(workerAddressSize)
+    buffer.position(buffer.position() + workerAddressSize)
+
+    val ep = clientConnections.getOrElseUpdate(id,
+      globalWorker.newEndpoint(new UcpEndpointParams().setUcpAddress(workerAddress))
+    )
+
+    val blockId: BlockId =
+      Utils.tryWithResource(new ByteBufferBackedInputStream(buffer))(bin => {
+        val in = new ObjectInputStream(bin)
+        in.readObject().asInstanceOf[BlockId]
+    })
+
+    val block = registeredBlocks.get(blockId)
+    if (block == null) {
+      throw new UcxException(s"Block $blockId not registered")
+    }
+    val lock = block.lock.readLock()
+    lock.lock()
+    val blockMemory = block.getMemoryBlock
+
+    ep.sendTaggedNonBlocking(blockMemory.address, blockMemory.size, tag, new UcxCallback {
+      override def onSuccess(request: UcpRequest): Unit = {
+        lock.unlock()
+      }
+
+      override def onError(ucsStatus: Int, errorMsg: String): Unit = {
+        logError(s"Failed to send $blockId: $errorMsg")
+        lock.unlock()
+      }
+    })
+  }
+
+  private def pinMemory(block: Block): UcpMemory = {
+    val blockMemory = block.getMemoryBlock
+    ucxContext.memoryMap(
+      memMapParams.setAddress(blockMemory.address).setLength(blockMemory.size))
+  }
+
+  /**
+   * Registers blocks using blockId on SERVER side.
+   */
+  override def register(blockId: BlockId, block: Block) {
+    val registeredBock: Block = if (ucxShuffleConf.pinMemory) {
+      UcxPinnedBlock(block, pinMemory(block))
+    } else {
+      block
+    }
+    registeredBlocks.put(blockId, registeredBock)
+  }
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  override def mutate(blockId: BlockId, block: Block, callback: OperationCallback): Unit = {
+    Future {
+      unregister(blockId)
+      register(blockId, block)
+    } andThen {
+      case Failure(t) => if (callback != null) {
+        callback.onComplete(new OperationResult {
+          override def getStatus: OperationStatus.Value = OperationStatus.FAILURE
+
+          override def getError: TransportError = new TransportError(t.getMessage)
+
+          override def getStats: Option[OperationStats] = None
+        })
+      }
+      case Success(_) => if (callback != null) {
+        callback.onComplete(new OperationResult {
+          override def getStatus: OperationStatus.Value = OperationStatus.SUCCESS
+
+          override def getError: TransportError = null
+
+          override def getStats: Option[OperationStats] = None
+        })
+      }
+    }
+  }
+
+  /**
+   * Indicate that this blockId is not needed any more by an application
+   */
+  override def unregister(blockId: BlockId): Unit = {
+    val block = registeredBlocks.remove(blockId)
+    if (block != null) {
+      block.lock.writeLock().lock()
```

Comment:
so this can create deadlocks right? 

UCX has no timeouts as far as I know, if I have a reader which took a `readLock` before, and now I try to get a `writeLock` in order to unregister it... the thread calling unregister is going to block forever. Please correct me if I am wrong.

I am OK if we deal with the locking later, but we should have a github issue or something along those lines that tracks what is going on.

Reply:
Why? `writeLock` is waiting for all read/write locks to be released. Read lock is released here: https://github.com/openucx/sparkucx/pull/25/files#diff-96099ead5b6c5311e53051819e90ed04R194

Reply:
Yes, we definitely call `unlock` I think in all cases from what I can tell, but happens if we have an outstanding send operation and the receiver dies?

As an aside, we should wrap all call in a try/finally if it could throw:

```
try {
  code that could throw
} finally {
  lock.unlock()
}
```